src/neural/loader.cc

```diff
@@ -46,25 +50,89 @@ FloatVectors LoadFloatsFromFile(const std::string& filename) {
     } else {
       bytes_read += sz;
       buffer.resize(bytes_read);
-      // Add newline in the end for the case it was not there.
-      buffer.push_back('\n');
       break;
     }
   }
   gzclose(file);
 
+  return buffer;
+}
+
+void DenormLayer(const lc0::Weights_Layer& layer, FloatVectors& vecs) {
+  FloatVector v;
+  auto& buffer = layer.params();
+  auto data = reinterpret_cast<const std::uint16_t*>(buffer.data());
+  int n = buffer.length() / 2;
+  v.resize(n);
+  for (int i = 0; i < n; i++) {
+    v[i] = data[i] / float(0xffff);
+    v[i] *= layer.max_val() - layer.min_val();
+    v[i] += layer.min_val();
+  }
+  vecs.emplace_back(v);
+}
+
+void DenormConvBlock(const lc0::Weights_ConvBlock& conv, FloatVectors& vecs) {
+  DenormLayer(conv.weights(), vecs);
+  DenormLayer(conv.biases(), vecs);
+  DenormLayer(conv.bn_means(), vecs);
+  DenormLayer(conv.bn_stddivs(), vecs);
+}
+
+FloatVectors LoadFloatsFromPbFile(const std::string& buffer) {
+  auto net = lc0::Net();
+  FloatVectors vecs;
+  net.ParseFromString(buffer);
+
+  std::string min_version(std::to_string(net.min_version().major()) + ".");
+  min_version += std::to_string(net.min_version().minor()) + ".";
+  min_version += std::to_string(net.min_version().patch());
+
+  if (net.min_version().major() > LC0_VERSION_MAJOR)
+    throw Exception("Weights require at least lc0 version: " + min_version);
+  if (net.min_version().minor() > LC0_VERSION_MINOR)
+    throw Exception("Weights require at least lc0 version: " + min_version);
+  if (net.min_version().patch() > LC0_VERSION_PATCH)
+    throw Exception("Weights require at least lc0 version: " + min_version);
+
+  if (net.format().weights_encoding() != lc0::Format::LINEAR16)
+    throw Exception("Invalid weight encoding");
+
+  auto& w = net.weights();
+
+  DenormConvBlock(w.input(), vecs);
```

Comment:
If `DenormConvBlock()` returned `FloatVector`, then I think
```cpp
  vecs.emplace_back(DenormConvBlock(w.input());
```
would move it magically without copying the memory and would be more readable.

Reply:
DenormConvBlock fills 4 layers of FloatVectors. So I don't see how this would work. DenormLayer() could certainly return a FloatVector.

Reply:
Indeed. Then just converting non-const reference parameter to pointer is what's needed.